scripts/state_space_models_lib.py

```diff
@@ -0,0 +1,316 @@
+                        elements: Tuple[Array, Array, Array, Array]):
+        mut_giv_T, Sigmat_giv_T = state
+        A = self.transition_matrix
+        mutt, Sigmatt, mut_cond_next, Sigmat_cond_next = elements
+        Jt  = Sigmatt @ A.T @ jnp.linalg.inv(Sigmat_cond_next)
+        mut_giv_T = mutt + Jt @ (mut_giv_T - mut_cond_next)
+        Sigmat_giv_T = Sigmatt + Jt @ (Sigmat_giv_T - Sigmat_cond_next) @ Jt.T
+        return (mut_giv_T, Sigmat_giv_T), (mut_giv_T, Sigmat_giv_T)
+    
+    def __forward_filter(self, x: Array) -> Tuple[Array, Array, Array, Array]:
+        mu0 = self.initial_state_prior.mean()
+        Sigma0 = self.initial_state_prior.covariance()
+        _, (filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist) = jax.lax.scan(self.__kalman_step, (mu0, Sigma0), x)
+        
+        return filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist
+    
+    def __backward_smoothing_pass(self, filtered_means: Array, filtered_covs: Array,
+                                  mu_cond_hist: Array, Sigma_cond_hist: Array) -> Tuple[Array, Array]:
+        # filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist = self.__forward_filter(x)
+        mut_giv_T = filtered_means[-1, :]
+        Sigmat_giv_T = filtered_covs[-1, :]
+
+        elements = (filtered_means[-2::-1], filtered_covs[-2::-1, ...], mu_cond_hist[1:][::-1, ...], Sigma_cond_hist[1:][::-1, ...])
+        _, (smoothed_means, smoothed_covs) = jax.lax.scan(self.__smoother_step, (mut_giv_T, Sigmat_giv_T), elements)
+        smoothed_means = jnp.concatenate([smoothed_means[::-1, ...], mut_giv_T[None, ...]], axis=0)
+        smoothed_covs = jnp.concatenate([smoothed_covs[::-1, ...], Sigmat_giv_T[None, ...]], axis=0)
+
+        return smoothed_means, smoothed_covs
+
+    def __sample_step(self, state_t: Array, key: PRNGKey):
+        key_state, key_obs = random.split(key)
+        state_mean_next = self.transition_matrix @ state_t
+
+        state_noise = self.transition_noise.sample(seed=key_state)
+        obs_noise = self.observation_noise.sample(seed=key_obs) 
+
+        state_next = state_mean_next + state_noise
+        obs_next = self.observation_matrix @ state_next + obs_noise
+
+        return state_next, (state_next, obs_next)
+    
+    def _sample_n(self, key: PRNGKey, n: int) -> Tuple[Array, Array]:
+        key_init, key_next, key_step = random.split(key, 3)
+        key_steps = random.split(key_step, n * self.num_timesteps).reshape(n, self.num_timesteps, -1)
+        
+        state_init = self.initial_state_prior.sample(seed=key_init, sample_shape=(n,))
+        
+        vmap_sample = jax.vmap(lambda states, keys: jax.lax.scan(self.__sample_step, states, keys))
+        _, (state_samples, obs_samples) = vmap_sample(state_init, key_steps)
+
+        return state_samples, obs_samples
+
+    def sample(self, *,
+               seed: Union[IntLike, PRNGKey],
+               sample_shape: Union[IntLike, Sequence[IntLike]] = ()) -> Array:
```

Comment:
you should pass num_timesteps as an optional argument, which defaults to self.max_num_timesteps if not specified. Then the event shape is (H*T, D*T) for hidden dim H, obs dim D, and num timesteps T